### PR TITLE
Assign HQ-defined roles to CCA user

### PR DESCRIPTION
**Pull request by Charl1996**
[Ticket](https://dimagi.atlassian.net/browse/SC-3473)

This PR expands on the DomainSyncUtil class's `sync_domain_role` method to now consider the roles for a particular user as defined on CommCare HQ. [This PR](https://github.com/dimagi/commcare-hq/pull/34987) is for returning the CCA roles for a particular user.

Note: This work will go [through QA](https://dimagi.atlassian.net/browse/QA-7055) and is already on staging.

### Approach
The approach is slightly different than the [spec](https://docs.google.com/document/d/1WTg8T37plRdjOK7wXl3-yip5SF_qG4LiD4hzMj57XOI/edit#heading=h.p76anzyemt26) suggests in the following ways:
1. We now create only 1 role for a user _per domain_ which has the correct permissions defined as per CommCare HQ. The role's name is composed [here](https://github.com/dimagi/commcare-analytics/blob/816f1f1735439eaf797021e25e6c58efcbdec726/hq_superset/utils.py#L289-L291) and is identifiable by domain and user.
2. When the user's permissions change on HQ and the user logs in again, the permissions for the role is overwritten with the new permissions.

By having only 1 role per domain user and setting the permissions for that role we won't end up having redundant roles that we'll need to clean up.

#### Side note
Since we're not assigning the `Gamma` by default to all users anymore (it's a configurable role on HQ now), it was found that the CCA user cannot log in to Superset since the privilege of having a profile on Superset is a permission the user needs.    If the user does not have this permission, the user's browser will redirect endlessly resulting in a browser error.

To account for this a new standard role, namely `hq_user`, is assigned to all CCA users, regardless of their permissions on CommCare HQ. This role has the necessary ("baseline") permissions needed for any users to be able have a profile on Superset/ be able to log in to Superset.

This means that users will need at least 3 distinct roles if they want to be able to access a domain's data on Superset:
1. HQ user role (new): The role that grants the user permission to be able to have a profile on Superset.
2. Domain schema role (existed before this PR): This role grants (and restricts) the user's access to a particular domain.
3. Domain user role (new): This role contains the necessary configurable permissions as defined on CommCare HQ and governs how the user can behave on the particular domain.

In addition to the above roles the user is able to have more roles if so defined on CommCare HQ. The roles available on HQ is currently `Gamma`, `dataset_editor` and `sql_lab`.

### CODE REVIEW

**Comment by mkangia** [comment]
Hey @Charl1996 

Can you link the related HQ PR in the description please?
https://github.com/dimagi/commcare-hq/pull/34987

**Comment by Charl1996** [comment]
> Hey @Charl1996
> 
> Can you link the related HQ PR in the description please? [dimagi/commcare-hq#34987](https://github.com/dimagi/commcare-hq/pull/34987)

Done

File: superset_config.example.py
Comment (mkangia) [inline superset_config.example.py:76]: Hey @Charl1996 

How do we roll this change out? Is there anything specifically needed for existing users? or their permission would reset correctly when they login.
So, will this role get removed from their account automatically now, if they had it earlier?

How does platform role work along with domain role? If someone gets the role sql_lab, it would be applicable only for the project they are currently on? and when they switch project, their platform roles are fetched and updated?

**Review by mkangia — COMMENTED** [review]
Beautiful PR @Charl1996 

I enjoyed reviewing this. :100: 

Are we taking this through QA?

**Comment by Charl1996** [comment]
> Are we taking this through QA?

Yes. I updated the description to make it more visible.

File: hq_superset/oauth.py
Comment (Charl1996) [reply hq_superset/oauth.py:59]: I was also thinking it feels a little out of place, but then all the permissions-related functionality sits in the class this class inherits from, and since this is a permissions-related method I thought it's probably best putting it here. Any other suggestions are welcome.

File: hq_superset/const.py
Comment (Charl1996) [reply hq_superset/const.py:46]: Correct. I played around a lot with how to best separate the read and write whilst keeping duplication to a minimum, but this current way of doing it feels like the most granular and configurable approach. You can set specific permissions for specific view/menu's

File: superset_config.example.py
Comment (Charl1996) [reply superset_config.example.py:76]: > or their permission would reset correctly when they login.

Yes, when they log in their permissions will reset correctly.

> So, will this role get removed from their account automatically now, if they had it earlier?

Correct. My thoughts are that if a user needs access they need to now have it explicitly granted on HQ (even if it means they might not have that permissions until it's explicitly given to them on HQ if they had it before), but maybe I can check with AEs as well.

> How does platform role work along with domain role? If someone gets the role sql_lab, it would be applicable only for the project they are currently on? 

Ah! Brilliant question! See response below.

> and when they switch project, their platform roles are fetched and updated?

More accurately, fetched and _overwritten_.

With the way it works in this PR, when a user logs in their user on CCA will be assigned roles/permissions based on only the domain's configuration they logged in to, so if their user on that domain on HQ does not have the Gamma role for instance, the user on CCA will not be granted that role when logging in. This is true even though the user might have Gamma on another domain; when they log into that other domain they will be given the Gamma role. So even though the platform role is technically platform wide, the user will only ever be able to have that role on the domain which gives it to the user. This is true because every time a user logs in we overwrite the user's roles completely with the appropriate roles for the domain the user logs into.

Comment (mkangia) [reply superset_config.example.py:76]: > Correct. My thoughts are that if a user needs access they need to now have it explicitly granted on HQ (even if it means they might not have that permissions until it's explicitly given to them on HQ if they had it before), but maybe I can check with AEs as well.

Right. It seems if we roll this out, people would lose access, at least temporarily, till they are assigned the permission on HQ.
The good thing is that the permissions setup is already available on HQ. So we can ask projects to set them even before we release this.

Additionally, we should be explicit that the permissions are updated only at login and when switching domains so they need to re-login to see the updated permissions from HQ.

Comment (Charl1996) [reply superset_config.example.py:76]: > permissions are updated only at login

This is not entirely true; the permissions are updated when selecting a domain, so even though the user is "logged in" to superset, when they change domains the permissions will update (sorry, maybe I wasn't too clear on this distinction in my previous comment)

Comment (mkangia) [reply superset_config.example.py:76]: Right, that is what I meant by "only at login and when switching domains"?

may be a correction to that "they need to re-login or re-select their domain" (I don't know if they can do that, if they have only one domain, which would be for most users?)

Comment (Charl1996) [reply superset_config.example.py:76]: Oh! Sorry, yeah, now I understand what you meant. You're right. Although there is a way when logged into the app to go to the "select domains" page without having to explicitly log out and re-auth.

Comment (Charl1996) [reply superset_config.example.py:76]: I've checked with AEs and it's OK that users might loose temporary access if their HQ permissions isn't set up correctly; we just have to make sure to communicate this change when deploying the new changes.

Comment (mkangia) [reply superset_config.example.py:76]: Cool.

**Review by mkangia — APPROVED** [review]
looking forward to QA on this.

File: hq_superset/hq_url.py
Comment (Charl1996) [inline hq_superset/hq_url.py]: HQ will now get the user from the request.

**Comment by Charl1996** [comment]
QA passed on this.

**Review by mkangia — APPROVED** [review]

**Review by kaapstorm — APPROVED** [review]

**Review by kaapstorm — APPROVED** [review]

**Review by mkangia — APPROVED** [review]

**Comment by Charl1996** [comment]
Please note that an [update](https://github.com/dimagi/commcare-analytics/pull/59/commits/cb57fa67dbc46e9b0bb9b8f9aeb52a3056914bc3) has been made which basically does the following:
1. If a HQ user has "edit" permissions set for CCA, assign the superset `Gamma` role
2. If a HQ user does not have "view" or "edit" permissions set for CCA, the user cannot access CCA.

File: hq_superset/tests/test_utils.py
Comment (mkangia) [inline hq_superset/tests/test_utils.py:45]: should not the user always have the custom domain role?

File: hq_superset/views.py
Comment (mkangia) [inline hq_superset/views.py:251]: nit: might be useful to add the name of the domain to the message so the user knows which domain they were trying to access?

Comment (Charl1996) [reply hq_superset/views.py:251]: Good idea!

File: hq_superset/tests/test_utils.py
Comment (Charl1996) [reply hq_superset/tests/test_utils.py:45]: Not anymore, since we assign the Gamma role to "edit" users and will only create and assign a "read only" role to "read" users. 

I think I'm going to update the custom user domain role name such that it makes clear that it's a read-only role. So the name becomes `<domain>_user_<user_id>_read_only`

File: hq_superset/views.py
Comment (Charl1996) [reply hq_superset/views.py]: https://github.com/dimagi/commcare-analytics/pull/59/commits/f5384d1e50aa6c3b11a8b6913838e4168984c643

File: hq_superset/tests/test_utils.py
Comment (Charl1996) [reply hq_superset/tests/test_utils.py:45]: https://github.com/dimagi/commcare-analytics/pull/59/commits/4d52e10f48937c255080e71371ec499e52c26bf1

Comment (mkangia) [reply hq_superset/tests/test_utils.py:45]: Oh you are just testing for the additional roles, sorry.

I was thinking about the domain role that is added to the user so they have access to just their domain?
This one, https://github.com/dimagi/commcare-analytics/blob/d8feb1d42c01b8b00ea8cc11bf81ae706c309d01/hq_superset/utils.py#L123

Comment (mkangia) [reply hq_superset/tests/test_utils.py:45]: > I think I'm going to update the custom user domain role name such that it makes clear that it's a read-only role. So the name becomes <domain>_user_<user_id>_read_only

Is that in line with the spec? We wanted to keep the naming consistent with roles. 
Additionally, this adds the extra overhead of changing the name to be without "read_only" if the permissions are updated, which is best avoided.

File: hq_superset/utils.py
Comment (mkangia) [inline hq_superset/utils.py:211]: should this be the one with "G"? The one on superset is "Gamma" or "gamma"?
Extracted it to a constant would be helpful to avoid confusion

File: hq_superset/utils.py
Comment (mkangia) [inline hq_superset/utils.py:293]: Hey @Charl1996 

I wanted to check with you if its possible to add the HQ role UUID to the name?
While looking at the ticket to expire permissions after every hour, I realized we currently don't a way of linking HQ role to the CCA role, which could be useful.
If its not a big lift, would be good to do it now since we are releasing this now.

File: hq_superset/tests/test_utils.py
Comment (Charl1996) [reply hq_superset/tests/test_utils.py:45]: The domain role that gives a user access to only their own domain is still given as usual and have nothing to do with these changes.

> Is that in line with the spec? We wanted to keep the naming consistent with roles.

Maybe not, but I thought it's probably a good idea to be able to see what permissions the role gives, since some users will have the `user_:user_id:_read_only` role and others not (I'm thinking in the scenario where someone's looking at a user's roles and want to know what access they have). If we give it a more generic name one would need to go and view the permissions individually to see what the role means.

Let me know if I need to change it back or if this new way is OK.

> Additionally, this adds the extra overhead of changing the name to be without "read_only" if the permissions are updated, which is best avoided.

It's just a one-liner, but yeah, that would need to update if we ever need to give more than read-only access that we decide to put into this role (as opposed to creating a different role).

File: hq_superset/utils.py
Comment (Charl1996) [reply hq_superset/utils.py:211]: > Extracted it to a constant would be helpful to avoid confusion

Good idea

File: hq_superset/utils.py
Comment (Charl1996) [reply hq_superset/utils.py:293]: I don't think it would be that much of a big lift. I'll look into it.

File: hq_superset/tests/test_utils.py
Comment (mkangia) [reply hq_superset/tests/test_utils.py:45]: Hey @Charl1996 

> If we give it a more generic name one would need to go and view the permissions individually to see what the role means.

So, the roles like dataset_editor are different right? This one is just for read only and write access? So the role would either be read only or write?

When a user's permissions are upgraded from read only to write, do we change the name of the same role or a new one gets created & old one deleted?

> Let me know if I need to change it back or if this new way is OK.

I think its fine to keep it if you think its going to be useful on the UI. I am just unsure of the  benefit over the cost of the overhead it brings, though we can always revisit this later if needed.

File: hq_superset/utils.py
Comment (mkangia) [reply hq_superset/utils.py:293]: Awesome, thank you!

File: hq_superset/tests/test_utils.py
Comment (mkangia) [reply hq_superset/tests/test_utils.py:45]: Actually,
Thinking more about this,
if the naming is updated to have `_read_only` in the role name, and if the users role is upgraded from view only to write, won't CCA end up having two roles when the syncing of role happens, since it will simply add a new role without the `_read_only` bit, leaving the old one around?

File: hq_superset/utils.py
Comment (Charl1996) [reply hq_superset/utils.py:294]: @mkangia I had a look into this and hit a snag...the Admin role is a virtual role and is never persisted in the database, so if the user is admin we don't have a role ID that we could use.

Maybe we should rethink how we track the role linkage. I'm going to hold off on this for now.

Comment (mkangia) [reply hq_superset/utils.py:294]: >  if the user is admin we don't have a role ID that we could use.

so the id could be "admin"?

>  I'm going to hold off on this for now.

We don't need to block on this. Fine to go ahead without the role id if needed.

**Comment by mkangia** [comment]
All looks good here except this point that I believe will add redundant roles in CCA
https://github.com/dimagi/commcare-analytics/pull/59#discussion_r1822857924

File: hq_superset/tests/test_utils.py
Comment (Charl1996) [reply hq_superset/tests/test_utils.py:45]: @mkangia 

> When a user's permissions are upgraded from read only to write, do we change the name of the same role or a new one gets created & old one deleted?

We don't assign the role and assign the Gamma instead (leaving the old one lying around).

Now that you mention this, I'm beginning to think we should simplify this to follow the previous way where we simply assigned a user domain role which will contain all the permissions (read or write). The user will always have this role, but the permissions will only be different. 

In the case of an edit user we can simply assign all the gamma permissions to the role. This is as easy as
```
new_permissions = gamma_role.permissions
```
(why did I not think of this earlier?)

What do you think?

Comment (mkangia) [reply hq_superset/tests/test_utils.py:45]: I'd suggest to use the Gamma role itself.

> assign all the gamma permissions to the role

This seems like duplication of permissions.
Additionally, Superset/Flask architecture supports adding multiple roles, so seems good to leverage a role that already exists with necessary permissions instead of copying its permissions over.

Though, effectively both approaches have the same end result, so seems okay to do this if you feel that is better.

Its additionally about if there is any difference of effort in these two
1. assign gamma role permissions
2. remove gamma role from user if the user does not have edit access

Comment (Charl1996) [reply hq_superset/tests/test_utils.py:45]: I'd like to add that for an "edit" user we assign the gamma role which also gives all the view permissions, so we don't assign the domain user role (the role that specifies which view/edit permissions the user has on a domain), meaning if a user was a "read only" user and is upgraded to the "edit" user there's a redundant view-only role in the DB that we'd need to clean up.

Comment (Charl1996) [reply hq_superset/tests/test_utils.py:45]: See [this PR](https://github.com/dimagi/commcare-analytics/pull/71) for alternative approach.

Comment (mkangia) [reply hq_superset/tests/test_utils.py:45]: > if a user was a "read only" user and is upgraded to the "edit" user there's a redundant view-only role in the DB that we'd need to clean up

Oh yes, I missed that.

**Comment by mkangia** [comment]
Hey @Charl1996 

Can you confirm what happens if HQ does not respond well or say responds with a 500 when fetching roles/permissions?
I had this happening locally and for some reason CCA continued to function which was unexpected.

**Comment by mkangia** [comment]
Hey @Charl1996 
Due to recent api changes that broke the integration, I was getting 500 from HQ, locally, that is where I noticed the pattern.
I see how this is handled now. I need to add the same here, https://github.com/dimagi/commcare-analytics/pull/72#discussion_r1828027612

I think the error message is a bit misleading thought. May be we could update to tell the user that we could not sync their permission instead of saying that they don't have permission.
Thank you.

**Comment by Charl1996** [comment]
If we merge [this PR](https://github.com/dimagi/commcare-hq/pull/35312) whilst keeping the resource on `v1` I'd need to just add the update here. I'll wait till it's merged before I do it.

File: hq_superset/utils.py
Comment (mkangia) [inline hq_superset/utils.py:162]: nit: I wonder if this is causing unnecessary saves on this role in the db since it would exist once it has been created.
Non blocking but good to check and do a follow up if needed.

I was thinking if this should instead be added by AUTH_USER_ADDITIONAL_ROLES instead though I am not sure how that would work since its a custom role and not predefined in superset.

File: hq_superset/utils.py
Comment (mkangia) [inline hq_superset/utils.py:170]: nit: should this check be on the top so the role isn't set if its already present?

File: hq_superset/utils.py
Comment (mkangia) [inline hq_superset/utils.py:237]: nit: do we really need these mapping? we can just return permissions and that should do right?

File: hq_superset/views.py
Comment (mkangia) [inline hq_superset/views.py:43]: just checking on why was this added?

File: hq_superset/views.py
Comment (mkangia) [inline hq_superset/views.py:69]: these checks were missing earlier?

**Review by mkangia — APPROVED** [review]
Hey @Charl1996 

Seems good to go ahead with this now.
 
Suggested some improvements, that are fine to taken as a follow up even as a separate ticket, if any of them seem reasonable. Considering the approach got simpler, which is great, code/implementation can be simplified further.

File: hq_superset/utils.py
Comment (Charl1996) [reply hq_superset/utils.py:139]: Ah, good catch!

**Comment by Charl1996** [comment]
@mkangia I'm going to merge this just to move it forward, but will be addressing your (very valid) comments in a followup PR.

File: hq_superset/views.py
Comment (Charl1996) [reply hq_superset/views.py:69]: Yes, but earlier we didn't need them. Now we need to set them to explicitly disable a read-only user from performing this action.

File: hq_superset/utils.py
Comment (Charl1996) [reply hq_superset/utils.py:237]: I guess you're right. Will update (in upcoming PR)

File: hq_superset/utils.py
Comment (Charl1996) [reply hq_superset/utils.py:162]: Mmm...I'll check it out

File: hq_superset/views.py
Comment (Charl1996) [reply hq_superset/views.py:43]: This was to ensure that only users with edit permissions can import data sources.

**Pull request merged**
